homework-submission/homeworkWeek1/main.jsx

```diff
@@ -0,0 +1,118 @@
+'use strict';
+
+const { Component } = React;
+
+// This component renders list items
+const ListItem = props => {
+  const { listData, isStatic, markAsCompleted } = props;
+  let { description, deadline, id, done } = listData;
+  return (
+    <li
+      id={id}
+      onClick={() => {
+        if (isStatic) {
+          const element = document.getElementById(id);
+          element.classList.toggle('done');
+        } else {
+          markAsCompleted(id - 1);
+        }
+      }}
+      className={done ? 'done' : 'not-done'}
+    >
+      {description},{deadline}
+    </li>
+  );
+};
+
+//This component creates a static list and displays listItem components
+const StaticList = props => {
+  const { isStatic } = props;
+  return (
+    <div>
+      <h2>Static List</h2>
+      <ul className="static list">
+        <ListItem
+          listData={{ id: 1, description: 'Get out of bed', deadline: '13.9.2017' }}
+          isStatic={isStatic}
+        />
+        <ListItem
+          listData={{ id: 2, description: 'Brush teeth', deadline: '14.9.2017' }}
+          isStatic={isStatic}
+        />
+        <ListItem
+          listData={{ id: 3, description: 'Eat breakfast', deadline: '15.9.2017' }}
+          isStatic={isStatic}
+        />
+      </ul>
+    </div>
+  );
+};
+
+//This component creates a dynamic list
+const DynamicList = props => {
+  const { listItems, markAsDone, isStatic } = props;
+  return (
+    <div>
+      <h2>Dynamic List</h2>
+      <ul className="dynamic list">
+        {listItems.map(listData => (
+          <ListItem
+            key={listData.id}
+            listData={listData}
+            markAsCompleted={markAsDone}
+            isStatic={isStatic}
+          />
+        ))}
+      </ul>
+    </div>
+  );
+};
+
+//This is the main component of the app act as a container
+class App extends Component {
+  state = {
+    dynamicList: [
+      {
+        id: 1,
+        description: 'Get out of bed',
+        deadline: '2017-09-11',
+        done: true,
+      },
+      {
+        id: 2,
+        description: 'Brush teeth',
+        deadline: '2017-09-10',
+        done: false,
+      },
+      {
+        id: 3,
+        description: 'Eat breakfast',
+        deadline: '2017-09-09',
+        done: false,
+      },
+    ],
+  };
+
+  markAsDone = index => {
+    let updateState = this.state.dynamicList[index];
+    updateState.done = !updateState.done;
+    this.setState(updateState);
```

Comment:
Now no need for this function, But I have a note on this, and it's important.

It's doing the job but it's not the very correct way, this is a very small application and it will not make any kind of affect here, but you will see the affect of that in a future applications.

Line 99 is just changing the state, but more accurately it's been changed before that.
Usually `setState` made for controlling the state in the correct order, so it made it's call and there is a complete lifecycle behind that call, it's not true it's only for rendering!

it have a lot happening behind, in a future lectures of course.
But just let's have our attention on line 98 which is changing the state, if you tried the following it will tell you that this line is changing it:
```javascript
    // if the first log is `true` the other `false` and this behavior tell it's been changed
    console.log(this.state.dynamicList[index].done)
    // this line is changing by reference not by value, which make the 
    updateState.done = !updateState.done;change on the stat by it self
    console.log(this.state.dynamicList[index])
```

perhaps this we will need to try changing the state in a more dynamic way, or keeping it immutable per say, so we could try:
```javascript
const updateState = this.state.dynamicList.map((item) => {
  if (item.id === this.state.dynamicList[index].id) {
    return {
      ...item, done: !item.done
    }
  } else {
    return item
  }
})
```

This is the solution for this thing, I have solved it, because it's a bit tricky and it will be removed either ways.
Please rethink about it, and if you have any question don't keep it, I'm here for help 🙂